c10/core/InferenceMode.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+
+namespace c10 {
+
+// A RAII, thread local (!) guard that enables or disables inference mode upon
+// construction, and sets it back to the original value upon destruction.
+struct TORCH_API InferenceMode {
+  InferenceMode(bool enabled=true): prev_keyset(c10::impl::tls_local_dispatch_key_set()) {
+    // Note [Expected TLS state in InferenceMode]:
+    //   InferenceMode: InplaceOrView not in included, Autograd in excluded
+    //   NormalMode: InplaceOrView in included, Autograd not in excluded
+    //
+    // 1. When InferenceMode is enabled, Autograd dispatch keys are excluded
+    //    but not InplaceOrView key.
+    //
+    //    For example:
+    //    torch::Tensor a = torch::ones({1, 2, 3}).set_requires_grad(true);
+    //    torch::Tensor k = a + 2;
+    //    {
+    //      c10::InferenceMode guard(true);
+    //      k.add_(2);
+    //    }
+    //    `k.add_(2)` still need to go through InplaceOrView kernel so that it's
+    //    prepared for future autograd.
+    //  2. When InferenceMode is disabled, InplaceOrView must be added
+    //     to included set.
+    //
+    //     For example:
+    //     torch::Tensor a;
+    //     {
+    //       c10::InferenceMode guard(true);
+    //       torch::Tensor in = torch::ones({2, 2});
+    //       a = in.view({1, 4});
+    //     }
+    //     torch::Tensor c = a.view({4, 1}); // (*)
+    //     If we don't add InplaceOrView to included set, (*) will skip its as_view
```

Comment:
Is this still true now that we have `c` that is also an inference Tensor?
Or is that change in another PR?

Reply:
Yea that's for the second PR. Will change it there!

Reply:
Yea that's for the second PR. Will change it there!